molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+                cpu,
+                'topology',
+                sockets=str(instance['cpu']['sockets']),
+                cores=str(instance['cpu']['cores']),
+                threads=str(instance['cpu']['threads']))
+        ET.SubElement(
+            dom, 'memory', unit='MiB').text = str(instance.get('memory', 1024))
+        os_element = ET.SubElement(dom, 'os')
+        ET.SubElement(os_element, 'type').text = 'hvm'
+        boot = ET.SubElement(os_element, 'boot', dev='hd')
+        ET.SubElement(dom, 'on_crash').text = 'restart'
+        features = ET.SubElement(dom, 'features')
+        for f in ['acpi', 'apic', 'pae']:
+            features.append(ET.Element(f))
+        devices = ET.SubElement(dom, 'devices')
+
+        # Disk elements
+        disk = ET.SubElement(devices, 'disk', type='file', device='disk')
+        ET.SubElement(
+            disk, 'driver', name='qemu', type='qcow2', cache='default')
+        ET.SubElement(
+            disk,
+            'source',
+            file=(os.path.join(self._pool_path, instance['name'] + '.img')))
+        ET.SubElement(disk, 'target', dev='vda', bus='virtio')
+
+        # Serial/console/usb elements
+        serial = ET.SubElement(devices, 'serial', type='pty')
+        target = ET.SubElement(serial, 'target', port='0')
+        alias = ET.SubElement(serial, 'alias', name='serial0')
+
+        console = ET.SubElement(devices, 'console', type='pty')
+        target = ET.SubElement(console, 'target', port='0', type='serial')
+        alias = ET.SubElement(console, 'alias', name='serial0')
+
+        # Network interface elements
+        #iface = ET.SubElement(devices, 'interface', type='network')
+        #ET.SubElement(iface, 'source', network='default')
+        for nic in instance.get('interfaces', [{'network_name': 'default'}]):
+            iface = ET.SubElement(devices, 'interface', type='network')
+            if 'address' in nic:
+                ip = netaddr.IPAddress(nic['address'])
+                # Find the libvirt network for this interface is connected to
+                cidr = netaddr.IPNetwork(
+                    filter(lambda net: net['name'] == nic['network_name'],
+                           self.molecule.config.config['libvirt']['networks'])[
+                               0]['cidr'])
+                ET.SubElement(
+                    iface,
+                    'ip',
+                    address=nic['address'],
+                    prefix=str(cidr.prefixlen))
+                ET.SubElement(
+                    iface,
+                    'route',
+                    family='ipv4',
+                    address=str(cidr.network),
+                    prefix=str(cidr.prefixlen),
+                    gateway=str(cidr.ip))
+            ET.SubElement(iface, 'source', network=nic['network_name'])
+            ET.SubElement(iface, 'model', type='virtio')
+        # Finally
+        domxml = ET.tostring(dom)
+        LOG.debug("\tXMLDesc: {}".format(domxml))
```

Comment:
To print stuff you are using
- `LOG`
- `utilities `
- `print`
- `pp`

I believe you need to settle on *one* and use it across the board.